docs/en/creating-widgets/supplemental.md

```diff
@@ -0,0 +1,869 @@
+			</button>
+		</div>
+	);
+});
+```
+
+**Class-based variant:**
+
+Here, both `myState` and `counter` are updated as part of the same application logic operation, so `@watch()` could be added to either or both of the fields, with the same net effect and performance profile in all cases:
+
+> src/widgets/MyEncapsulatedStateWidget.tsx
+
+```tsx
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+import watch from '@dojo/framework/core/decorators/watch';
+import { tsx } from '@dojo/framework/core/vdom';
+
+export default class MyEncapsulatedStateWidget extends WidgetBase {
+	private myState: string = 'Hello from a stateful widget!';
+
+	@watch() private counter: number = 0;
+
+	protected render() {
+		return (
+			<div>
+				Current widget state: {this.myState}
+				<br />
+				<button
+					onclick={() => {
+						this.myState = 'State change iteration #' + ++this.counter;
+					}}
+				>
+					Change State
+				</button>
+			</div>
+		);
+	}
+}
+```
+
+## Intermediate: passing widget properties
+
+Passing state into a widget via virtual node [`properties`](#node-properties) is the most effective way of wiring up reactive data flows within a Dojo application.
+
+Widgets specify their own properties [interface](https://www.typescriptlang.org/docs/handbook/interfaces.html) which can include any fields the widget wants to publicly advertise to consumers, including configuration options, fields representing injectable state, as well as any event handler functions.
+
+Function-based widgets pass their properties interface as a generic type argument to the `create().properties()` call. The factory returned from this call chain then makes property values available via a `properties` argument in the render function definition.
+
+Class-based widgets can define their properties interface as a generic type argument to `WidgetBase` in their class definition.
+
+For example, a widget supporting state and event handler properties:
+
+> src/widgets/MyWidget.tsx
+
+**Function-based variant:**
+
+```tsx
+import { create, tsx } from '@dojo/framework/core/vdom';
+
+const factory = create().properties<{
+	name: string;
+	onNameChange?(newName: string): void;
+}>();
+
+export default factory(function MyWidget({ properties: { name, onNameChange } }) {
+	let newName: string = '';
+
+	return (
+		<div>
+			<span>Hello, {name}! Not you? Set your name:</span>
+			<input
+				type="text"
+				oninput={(e: Event) => {
+					newName = (e.target as HTMLInputElement).value;
+				}}
+			/>
+			<button
+				onclick={() => {
+					onNameChange && onNameChange(newName);
+				}}
+			>
+				Set new name
+			</button>
+		</div>
+	);
+});
+```
+
+**Class-based variant:**
+
+```tsx
+import WidgetBase from '@dojo/framework/core/WidgetBase';
+import { tsx } from '@dojo/framework/core/vdom';
+
+export interface MyWidgetProperties {
+	name: string;
+	onNameChange?(newName: string): void;
+}
+
+export default class MyWidget extends WidgetBase<MyWidgetProperties> {
```

Comment:
@sbinge I think this should use widget state for the input name and set the input value on each render, something like this:

```tsx
export default class MyWidget extends WidgetBase<MyWidgetProperties> {
	private newName = '';
	protected render() {
		const { name, onNameChange } = this.properties;
		return (
			<div>
				<span>Hello, {name}! Not you? Set your name:</span>
				<input
					type="text"
					value={this.newName}
					oninput={(e: Event) => {
						this.newName = (e.target as HTMLInputElement).value;
						this.invalidate();
					}}
				/>
				<button
					onclick={() => {
						this.newName = '';
						onNameChange && onNameChange(newName);
					}}
				>
					Set new name
				</button>
			</div>
		);
	}
}
```